Assignment4/rubykoom-21C008/task07.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+"""Task07.ipynb
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/1OqSG5GBfs1jJvFTF380Z3kez9dkDEQG-
+
+**Task 07: Querying RDF(s)**
+"""
+
+github_storage = "https://raw.githubusercontent.com/FacultadInformatica-LinkedData/Curso2022-2023/master/Assignment4/course_materials"
+
+"""Leemos el fichero RDF de la forma que lo hemos venido haciendo"""
+
+from rdflib import Graph, Namespace, Literal
+from rdflib.namespace import RDF, RDFS
+from rdflib.plugins.sparql import prepareQuery
+g = Graph()
+g.namespace_manager.bind('ns', Namespace("http://somewhere#"), override=False)
+g.namespace_manager.bind('vcard', Namespace("http://www.w3.org/2001/vcard-rdf/3.0#"), override=False)
+g.parse(github_storage+"/rdf/example6.rdf", format="xml")
+
+"""**TASK 7.1: List all subclasses of "Person" with RDFLib and SPARQL**"""
+
+# TO DO
+ns = Namespace("http://somewhere#")
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+
+# RDFlib:
+subclass_of_person = []   # Create an empty list where store requested data
+print("Results RDFlib:")
+print("")
+for s, p, o in g.triples((None, RDFS.subClassOf, ns.Person)): # Accomplish this
+    print(s)
+    subclass_of_person.append(s)   # Adding them to empty list given
+
+# SPARQL:
+print("")
+print("Results SPARQL:")
+print("")
+query1 = prepareQuery('''
+  SELECT ?Subclass WHERE { 
+    ?Subclass RDFS:subClassOf ns:Person.  
+  }
+  ''',
+                  initNs={"RDFS": RDFS, "ns": ns}
+                  )
+
+# Visualize the results
+for r in g.query(query1):
+    print(r.Subclass)
+
+"""**TASK 7.2: List all individuals of "Person" with RDFLib and SPARQL (remember the subClasses)**
+
+"""
+
+# TO DO
+
+# RDFlib:
+people = []  # Create an empty list where store requested data.
+print("Results RDFlib:")
+print("")
+for s, p, o in g.triples((None, RDF.type, ns.Person)):  # Individuals who belong to class Person.
+    people.append(s)
+    print(s)
+for subclass in subclass_of_person:
+    for s1, p1, o1 in g.triples((None, RDF.type, subclass)): # Individuals who belong to sublcass Person.
+        print(s1)
+        people.append(s1)
+
+# SPARQL:
+print("")
+query2 = prepareQuery('''
+  SELECT DISTINCT ?People WHERE {
+    {
+      ?People RDF:type ns:Person.             # Individuals who belong to class Person.
+    }
+    UNION
+    {
+      ?Subclass RDFS:subClassOf ns:Person.
```

Comment:
This is nice, but if you wanted to get ALL subclasses and their respective subclasses, it would be better to use subClassOf*. That way if you have that Student is a subclass of Researcher that is subclass of Person, you would also obtain Student among the results

Reply:
Done. As you have said, by this way you obtain subclass PhDStudent of subclass Researcher.